src/lib.rs

```diff
@@ -1730,7 +1730,31 @@ pub trait Itertools : Iterator {
         }
         None
     }
+    /// Find the value of the first element satisfying a predicate or return the last element, if any.
+    ///
+    /// The iterator is not advanced past the first element found.
+    ///
+    /// ```
+    /// use itertools::Itertools;
+    ///
+    /// let numbers = [1, 2, 3, 4];
+    /// assert_eq!(numbers.iter().find_or_last(|&&x| x > 5), Some(&4));
+    /// assert_eq!(numbers.iter().find_or_last(|&&x| x > 2), Some(&3));
+    /// assert_eq!(std::iter::empty::<i32>().find_or_last(|&x| x > 5), None);
+    /// ```
+	fn find_or_last<P>(mut self, predicate: P) -> Option<Self::Item>
+		where Self: Sized,
+			  P: FnMut(&Self::Item) -> bool,
+	{
+		#[inline]
+        fn check<T>(mut predicate: impl FnMut(&T) -> bool) -> impl FnMut(Option<T>, T) -> Result<Option<T>, T> {
```

Comment:
I've seen the pattern of `inline`d local functions sometimes, but I do not know if we use it regularly in `Itertools`. Could you possibly shed some light on the benefit of this trick? Does it generate more efficient code?

Reply:
What it's trying to do is reduce the number of monomorphizations that happen, and thus be a bit faster to compile.  A closure would monomorphize for every iterator type *and* every predicate type, whereas the local function just monomorphizes for the latter.  That said, I'm not sure how essential that is here.  (It makes more sense inside `try_fold`, where predicates like "always returns `Ok`" are more plausible to see in multiple places times.)

Reply:
It seems to be that way ([inlining docs](https://nnethercote.github.io/perf-book/inlining.html)). The way I came up with this implementation is by basically combining [Iterator::last](https://doc.rust-lang.org/src/core/iter/traits/iterator.rs.html#273-275) and [Iterator::find](https://doc.rust-lang.org/src/core/iter/traits/iterator.rs.html#2323-2326), where they used `#[inline]`.

Reply:
@scottmcm [try_fold](https://doc.rust-lang.org/src/core/iter/traits/iterator.rs.html#1973-1986) only uses the provided predicate, I am not sure what you mean with using `#[inline]` inside `try_fold`.

Reply:
See https://github.com/rust-lang/rust/pull/62429 for why the ones in `core` are written like that.